### PR TITLE
Length of sentence check fix

### PR DESCRIPTION
**Pull request by tomy983**
Increased _NMEA_MIN_LENGTH, as it is pointless to send to NTRIP server a string without position info like this:  $GNGGA,160653.60,,,,,0,00,99.99,,,,,,*79
Also made sure that maximum and minimum sentence check is performed excluding \r or \n

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,5 +1,11 @@
 # ROS NTRIP Client
 
+Cloned from https://github.com/LORD-MicroStrain/ntrip_client with the addition of https://github.com/PaulBouchier/ntrip_client/tree/rtcm_msgs_support for rtcm_msgs/Message support. I am using it with https://github.com/KumarRobotics/ublox and a ZED-F9P on a raspberry pi 4 running ROS noetic.
```
Comment (robbiefish) [inline README.md]: This section should be removed before merge

File: launch/ntrip_client.launch [hunk 2]
```diff
@@ -4,14 +4,15 @@
...
-  <arg name="mountpoint"             default = "VTRI_RTCM3" />
+  <arg name="host"                   default = "IP_ADD" />
+  <arg name="port"                   default = "PORT" />
+  <arg name="mountpoint"             default = "MNT_POINT" />
   <arg name="authenticate"           default = "true" />
   <arg name="username"               default = "user" />
   <arg name="password"               default = "pass" />
   <arg name="ntrip_version"          default = "" />
   <arg name="debug"                  default = "false" />
+  <arg name="rtcm_msg_pkg"           default = "rtcm_msgs" />
```
Comment (robbiefish) [inline launch/ntrip_client.launch]: The default should still be `mavros_msgs`

File: scripts/ntrip_ros.py [hunk 3]
```diff
@@ -102,15 +126,27 @@ def subscribe_nmea(self, nmea):
...
+          rtcm_msg = Message(
+            header=Header(
+              stamp=rospy.Time.now(),
+              frame_id=self._rtcm_frame_id
+            ),
+            message =raw_rtcm
+          )
+        self._rtcm_pub.publish(rtcm_msg)
+    except:
+      pass
```
Comment (robbiefish) [inline scripts/ntrip_ros.py]: Would probably be good to log the error here. That can get a little noisy, but failing silently may cause confusion

File: src/ntrip_client/nmea_parser.py [hunk 4]
```diff
@@ -15,11 +15,11 @@ def __init__(self, logerr=logging.error, logwarn=logging.warning, loginfo=loggin
 
   def is_valid_sentence(self, sentence):
     # Simple sanity checks
-    if len(sentence) > _NMEA_MAX_LENGTH:
+    if len(sentence.rstrip()) > _NMEA_MAX_LENGTH:
```
Comment (robbiefish) [inline src/ntrip_client/nmea_parser.py]: All documentation I can find on the NMEA spec says that the max length for a sentence is 82 characters including the `$` and `\r\n`. If you need the parser to support longer sentences, perhaps this should be a parameter to the `NMEAParser`.

See:
https://gpsd.gitlab.io/gpsd/NMEA.html
https://en.wikipedia.org/wiki/NMEA_0183

File: src/ntrip_client/nmea_parser.py [hunk 5]
```diff
@@ -15,11 +15,11 @@ def __init__(self, logerr=logging.error, logwarn=logging.warning, loginfo=loggin
...
       self._logwarn('Received invalid NMEA sentence. Max length is {}, but sentence was {} bytes'.format(_NMEA_MAX_LENGTH, len(sentence)))
       self._logwarn('Sentence: {}'.format(sentence))
       return False
-    if len(sentence) < _NMEA_MIN_LENGTH:
+    if len(sentence.rstrip()) < _NMEA_MIN_LENGTH:
```
Comment (robbiefish) [inline src/ntrip_client/nmea_parser.py]: Same as above comment

File: src/ntrip_client/ntrip_client.py [hunk 6]
```diff
@@ -23,6 +24,13 @@
 
 class NTRIPClient:
 
+  _basic_credentials = None
+  _connected = False
+  _valid_nmea = False
+  _lastNEMEA = None
```
Comment (robbiefish) [inline src/ntrip_client/ntrip_client.py]: These variables should be snake case, and `NEMEA` should be `NMEA` unless `NEMEA` is an abbreviation I am unaware of, for example: `_last_nmea`

File: src/ntrip_client/ntrip_client.py [hunk 7]
```diff
@@ -139,38 +172,71 @@ def send_nmea(self, sentence):
...
-      return []
-
-    # Check if there is any data available on the socket
-    read_sockets, _, _ = select.select([self._server_socket], [], [], 0)
-    if not read_sockets:
-      return []
-
     # Since we only ever pass the server socket to the list of read sockets, we can just read from that
     # Read all available data into a buffer
+    lastread = 0
```
Comment (robbiefish) [inline src/ntrip_client/ntrip_client.py]: snake case

**Review by robbiefish — CHANGES_REQUESTED** [review]
Thanks so much for contributing these changes! I will test them out ASAP, and get back with some more comments. In the meantime, there are a couple of changes to code style and other things I would like to see changed before this is merged

File: README.md (hunk 1 above)
Comment (tomy983) [reply README.md]:
```suggestion
Also compatible with https://github.com/KumarRobotics/ublox and  ZED-F9P 
```

File: launch/ntrip_client.launch (hunk 2 above)
Comment (tomy983) [reply launch/ntrip_client.launch]:
```suggestion
  <arg name="rtcm_msg_pkg"           default = "mavros_msgs" />
```

File: src/ntrip_client/nmea_parser.py (hunk 4 above)
Comment (tomy983) [reply src/ntrip_client/nmea_parser.py]: The standard is 82 characters as you say. The ZED-F9P, in high precision mode can actually have more (i am struggling to find out how many more, but never seen more than 82 EXCLUDING \r\n...) see page 7 first table https://cdn.sparkfun.com/assets/f/7/4/3/5/PM-15136.pdf  . Also other vendors have something similar. 
It would be a shame to have the high precision mode truncated if we want to use the same string for other porpoises (more than enough for the ntrip server though...)

File: src/ntrip_client/nmea_parser.py (hunk 5 above)
Comment (tomy983) [reply src/ntrip_client/nmea_parser.py]: Also the same... I did't know about the 82 characters max length so I thought it was a mistake not including those endline  characters... I think the best would be to just increase the max length to something more than 85ish and remove my changes...

File: src/ntrip_client/ntrip_client.py (hunk 6 above)
Comment (tomy983) [reply src/ntrip_client/ntrip_client.py]: yep :-)

File: src/ntrip_client/ntrip_client.py (hunk 7 above)
Comment (tomy983) [reply src/ntrip_client/ntrip_client.py]: also yes

**Comment by tomy983** [comment]
Most of the changes have been made as requested. Some of those I am not really sure how to address them (for example the max nmea length...) Maybe adding a parameter... But this would probably generate more confusion than is worth.

File: src/ntrip_client/nmea_parser.py (hunk 4 above)
Comment (robbiefish) [reply src/ntrip_client/nmea_parser.py]: I see, that use case makes sense. However, I am hesitant to just increase the limit. In this case, I think the most flexible thing to do would be to log a debug message when we receive a NMEA message that is longer than 82 characters but is still valid otherwise. The message could be something like:
```
Received NMEA sentence longer than 82 characters that was otherwise valid. This message will still be considered valid to support fine precision NMEA sentences
```

In this case, the `.rstrip()` should be removed on this line and the following instance, and the `_NMEA_MAX_LENGTH` should be changed back to 82

Comment (tomy983) [reply src/ntrip_client/nmea_parser.py]: Yes, I agree on removing `.rstrip()` . I already did. (see latest commits). I like your idea, but was wondering if instead of trowing a debug message (somewhat hidden), what if we just leave the limit at 82 and log an error message explaining what might be going on and how it can be fixed? In such case we could also use a parameter and leaving it "hidden" unless required.. Just an idea, whatever you like the most. 
The message, in such case could be something like:
`Received NMEA sentence longer than 82 characters. If an high precision device is in use, the NMEA length could be longer. In such case add the following parameter to your launch file, specifying the max possible string length: <param name="nmea_max_length"  value="" />`

Comment (robbiefish) [reply src/ntrip_client/nmea_parser.py]: Yeah I like that solution, we could default the max length to 82 if `nmea_max_length` is not set, otherwise use whatever that is set to, and then in the launch file default `nmea_max_length` to 82 as well and document why someone might change the value

File: src/ntrip_client/ntrip_client.py [hunk 8]
```diff
@@ -23,6 +24,13 @@
 
 class NTRIPClient:
 
+  _basic_credentials = None
+  _connected = False
+  _valid_nmea = False
+  _last_nmea = None
+  _justConnected = False
```
Comment (robbiefish) [inline src/ntrip_client/ntrip_client.py]: Looks like there are still some camel case variables that should be converted to snake case (same on line below)

**Comment by tomy983** [comment]
Hi, I should have the variables names cleaned now.
I also, as we discussed, added the option for the maximum nmea length [#](https://github.com/LORD-MicroStrain/ntrip_client/pull/13#discussion_r889014633) I've tested the code and (for me) is working no problem). I am throttling the nmea messages coming in from the gps as not to flood the ntrip server, limit data usage and also limit warnings log in case of bad nmea strings. 
Apart from this, It's working alright.

File: src/ntrip_client/nmea_parser.py (hunk 4 above)
Comment (tomy983) [reply src/ntrip_client/nmea_parser.py]: Done

**Comment by robbiefish** [comment]
Apologies that I have left this open for so long. I finally got the chance to test things out, and the NMEA sentence length change appears to work as expected. However, the reconnect logic does not appear to be working properly. When using a network (virtual) mountpoint, the driver will keep attempting to reconnect until the user sends a NMEA sentence

I think that the right thing to do here is to either update this PR to do just what the title suggests (adjustable NMEA length) or make a new PR with just the NMEA sentence length changes that way we can get it merged in without being held up on the reconnect issues.

I also think that the message package switching is implemented properly but should also probably be split into it's own PR. If you don't have the time, I am happy to make the PRs for you, just want to give you the chance to get the Github credit :)

Also, I took a look at the connection logic, and adapted your code to a different branch. If possible, please let me know if this reconnect logic also solves the reconnect problem for you: [ROS](https://github.com/LORD-MicroStrain/ntrip_client/tree/feature/ros_reconnect), [ROS2](https://github.com/LORD-MicroStrain/ntrip_client/tree/feature/ros2_reconnect)

**Comment by tomy983** [comment]
> Apologies that I have left this open for so long. I finally got the chance to test things out, and the NMEA sentence length change appears to work as expected. However, the reconnect logic does not appear to be working properly. When using a network (virtual) mountpoint, the driver will keep attempting to reconnect until the user sends a NMEA sentence
> 
IIRC it tries to reconnect every time a nmea message to be sent to the ntrip arrives on his topic. Id does not happens to me because I throttle the nmea topic like this:
`<node name="nmea_throttler" type="throttle" pkg="topic_tools" args="messages /gps_link/nmea 0.1 /ntrip_client/nmea" />`
I do this for two reasons:
- To avoid flooding the ntrip server with my position (some servers do not react well and some of them specifically ask for updates at least every n seconds. I read about a german user on a public ntrip who had complains for too frequent nmea updates. Can't find the reference now..)
- To save on data transfer (eventually I will connect via radio modem)
I don't think there is any points in sending the nmea to the ntrip server so often (in my case it would otherwise be at 5Hz), so I throttle the message to publish every 10 seconds which for me is good enough for both update my position to the ntrip and also keeping the connection with it alive.
When a new nmea message is received it is sent to the ntrip server. If the connection is lost it will attempt to reconnect. 
This (for me) also has the advantage that when testing inside and no valid nmea sentence is received ([minimum lenght...](https://github.com/tomy983/ntrip_client/blob/468367a6f28df1b18134c9df7473f3f5e24f7094/src/ntrip_client/nmea_parser.py#L3)) my ntrip server after about 60 sec disconnects me, but I will not attempt a reconnection until a valid nmea is published. When I move outside, a valid sentence is received and it will attempt a reconnection (every 10 seconds)

To recap: is the frequency of a valid nmea message that set the reconnection timing.
Btw, this would probably be a nice option to have in your code, instead of having to use topic_tools..

>I think that the right thing to do here is to either update this PR to do just what the title suggests (adjustable NMEA length) or make a new PR with just the NMEA sentence length changes that way we can get it merged in without being held up on the reconnect issues.
>
It is not much code around that, would you mind just copy the part that you need?
> I also think that the message package switching is implemented properly but should also probably be split into it's own PR. If you don't have the time, I am happy to make the PRs for you, just want to give you the chance to get the Github credit :)
> 
That part of the code that deals with this is taken from [here](https://github.com/PaulBouchier/ntrip_client/tree/rtcm_msgs_support) and you could create a pr from it.
Thank you for the consideration. I am grateful I have found your ntrip_client, @PaulBouchier adaptation was really useful so I used it and I just fixed what did not work for me. You absolutely go ahead and use whatever part you find useful as you wish.

> Also, I took a look at the connection logic, and adapted your code to a different branch. If possible, please let me know if this reconnect logic also solves the reconnect problem for you: [ROS](https://github.com/LORD-MicroStrain/ntrip_client/tree/feature/ros_reconnect), [ROS2](https://github.com/LORD-MicroStrain/ntrip_client/tree/feature/ros2_reconnect)

Ok, I will give it a try. 
I think we can close this pr.

**Comment by robbiefish** [comment]
Closing in favor of the following PRs:

* NMEA sentence length configuration:
    * ROS: #19 
    * ROS2: #20
* Different RTCM message publishing:
    * ROS: #22 
    * ROS2: #21
* Automatic reconnect:
    * ROS: #18 
    * ROS2: #17
